README.md
# story_painter
This package provides drawing experince like in story editor of Instagram.

**story_painter** is a modified version of the [hand_signature](https://pub.dev/packages/hand_signature).

<img src="https://firebasestorage.googleapis.com/v0/b/app-monotony.appspot.com/o/assets%2Fdemo.gif?alt=media&token=2e525617-acb6-40f0-bd68-f3e72c0e97c6"/>

### Check out **fast_color_picker** from [here](https://pub.dev/packages/fast_color_picker). 

## Usage

Initialize the controller

``` Dart
StoryPainterControl painterControl = StoryPainterControl(
    type: PainterDrawType.shape,
    threshold: 3.0,
    smoothRatio: 0.65,
    velocityRange: 2.0,
    color: Colors.white,
    width: 8,
    onDrawStart: () {},
    onDrawEnd: () {},
);
```

Build StoryPainter with controller

``` Dart
StoryPainter(control: painterControl)
```

Change brush specs while drawing

``` Dart
painterControl.setColor(Colors.red);
painterControl.setWidth(24.0);
```

Export the image

``` Dart
ui.Image image = await painterControl.toImage(pixelRatio: 3.0);
```

Display or use the output

``` Dart
// Convert the data
ByteData byteData = await image.toByteData(format: ui.ImageByteFormat.png);
Uint8List pngBytes = byteData.buffer.asUint8List();

// Display with memory image
Image.memory(pngBytes)
```



CHANGELOG.md
## [0.0.4]
* dependency upgrade
## [0.0.3]
* nullsafety migration.
## [0.0.2]
* initialization.


test/story_painter_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:story_painter/story_painter.dart';

void main() {
  test('adds one to input values', () {});
}


lib/src/utils.dart
import 'dart:math';
import 'dart:ui';

import 'package:flutter/cupertino.dart';
import 'package:flutter_svg/flutter_svg.dart';

import '../story_painter.dart';

extension ColorEx on Color {
  String get hexValue => '#${value.toRadixString(16)}'.replaceRange(1, 3, '');
}

extension OffsetEx on Offset {
  Offset axisDistanceTo(Offset other) => other - this;

  double distanceTo(Offset other) {
    final len = axisDistanceTo(other);

    return sqrt(len.dx * len.dx + len.dy * len.dy);
  }

  double angleTo(Offset other) {
    final len = axisDistanceTo(other);

    return atan2(len.dy, len.dx);
  }

  Offset directionTo(Offset other) {
    final len = axisDistanceTo(other);
    final m = sqrt(len.dx * len.dx + len.dy * len.dy);

    return Offset(len.dx / m, len.dy / m);
  }

  Offset rotate(double radians) {
    final s = sin(radians);
    final c = cos(radians);

    final x = dx * c - dy * s;
    final y = dx * s + dy * c;

    return Offset(x, y);
  }

  Offset rotateAround(Offset center, double radians) {
    return (this - center).rotate(radians) + center;
  }
}

extension PathEx on Path {
  void start(Offset offset) => moveTo(offset.dx, offset.dy);

  void cubic(Offset cpStart, Offset cpEnd, Offset end) =>
      cubicTo(cpStart.dx, cpStart.dy, cpEnd.dx, cpEnd.dy, end.dx, end.dy);

  void line(Offset offset) => lineTo(offset.dx, offset.dy);
}

//TODO: clean up
class PathUtil {
  static Rect bounds(List<Offset> data) {
    double left = data[0].dx;
    double top = data[0].dy;
    double right = data[0].dx;
    double bottom = data[0].dy;

    data.forEach((point) {
      final x = point.dx;
      final y = point.dy;

      if (x < left) {
        left = x;
      } else if (x > right) {
        right = x;
      }

      if (y < top) {
        top = y;
      } else if (y > bottom) {
        bottom = y;
      }
    });

    return Rect.fromLTRB(left, top, right, bottom);
  }

  static Rect boundsOf(List<List<Offset>> data) {
    double left = data[0][0].dx;
    double top = data[0][0].dy;
    double right = data[0][0].dx;
    double bottom = data[0][0].dy;

    data.forEach((set) => set.forEach((point) {
          final x = point.dx;
          final y = point.dy;

          if (x < left) {
            left = x;
          } else if (x > right) {
            right = x;
          }

          if (y < top) {
            top = y;
          } else if (y > bottom) {
            bottom = y;
          }
        }));

    return Rect.fromLTRB(left, top, right, bottom);
  }

  static List<T> translate<T extends Offset>(List<T> data, Offset location) {
    final output = <T>[];

    data.forEach(
        (point) => output.add(point.translate(location.dx, location.dy) as T));

    return output;
  }

  static List<List<T>> translateData<T extends Offset>(
      List<List<T>> data, Offset location) {
    final output = <List<T>>[];

    data.forEach((set) => output.add(translate(set, location)));

    return output;
  }

  static List<T> scale<T extends Offset>(List<T> data, double ratio) {
    final output = <T>[];

    data.forEach((point) => output.add(point.scale(ratio, ratio) as T));

    return output;
  }

  static List<List<T>> scaleData<T extends Offset>(
      List<List<T>> data, double ratio) {
    final output = <List<T>>[];

    data.forEach((set) => output.add(scale(set, ratio)));

    return output;
  }

  static List<T> normalize<T extends Offset>(List<T> data,
      {Rect? bound, double? border}) {
    bound ??= bounds(data);
    border ??= 0.0;

    return scale<T>(
      translate<T>(data, -bound.topLeft + Offset(border, border)),
      1.0 / (max(bound.width, bound.height) + border * 2.0),
    );
  }

  static List<List<T>> normalizeData<T extends Offset>(List<List<T>> data,
      {Rect? bound}) {
    bound ??= boundsOf(data);

    final ratio = 1.0 / max(bound.width, bound.height);

    return scaleData<T>(
      translateData<T>(data, -bound.topLeft),
      ratio,
    );
  }

  static List<T> fill<T extends Offset>(List<T> data, Rect rect,
      {Rect? bound, double? border}) {
    bound ??= bounds(data);
    border ??= 4.0;

    final outputSize = rect.size;
    final sourceSize = bound;
    Size destinationSize;

    if (outputSize.width / outputSize.height >
        sourceSize.width / sourceSize.height) {
      destinationSize = Size(
          sourceSize.width * outputSize.height / sourceSize.height,
          outputSize.height);
    } else {
      destinationSize = Size(outputSize.width,
          sourceSize.height * outputSize.width / sourceSize.width);
    }

    destinationSize = Size(destinationSize.width - border * 2.0,
        destinationSize.height - border * 2.0);
    final borderSize = Offset(rect.width - destinationSize.width,
            rect.height - destinationSize.height) *
        0.5;

    return translate<T>(
        scale<T>(
          normalize<T>(data, bound: bound),
          max(destinationSize.width, destinationSize.height),
        ),
        borderSize);
  }

  static List<List<T>> fillData<T extends Offset>(List<List<T>> data, Rect rect,
      {Rect? bound, double? border}) {
    bound ??= boundsOf(data);
    border ??= 4.0;

    final outputSize = rect.size;
    final sourceSize = bound;
    Size destinationSize;

    if (outputSize.width / outputSize.height >
        sourceSize.width / sourceSize.height) {
      destinationSize = Size(
          sourceSize.width * outputSize.height / sourceSize.height,
          outputSize.height);
    } else {
      destinationSize = Size(outputSize.width,
          sourceSize.height * outputSize.width / sourceSize.width);
    }

    destinationSize = Size(destinationSize.width - border * 2.0,
        destinationSize.height - border * 2.0);
    final borderSize = Offset(rect.width - destinationSize.width,
            rect.height - destinationSize.height) *
        0.5;

    return translateData<T>(
        scaleData<T>(
          normalizeData<T>(data, bound: bound),
          max(destinationSize.width, destinationSize.height),
        ),
        borderSize);
  }

  static Path toPath(List<Offset> points) {
    final path = Path();

    if (points.length > 0) {
      path.moveTo(points[0].dx, points[0].dy);
      points.forEach((point) => path.lineTo(point.dx, point.dy));
    }

    return path;
  }

  static List<Path> toPaths(List<List<Offset>> data) {
    final paths = <Path>[];

    data.forEach((line) => paths.add(toPath(line)));

    return paths;
  }

  static Rect pathBounds(List<Path> data) {
    Rect init = data[0].getBounds();

    double left = init.left;
    double top = init.top;
    double right = init.right;
    double bottom = init.bottom;

    data.forEach((path) {
      final bound = path.getBounds();

      left = min(left, bound.left);
      top = min(top, bound.top);
      right = max(right, bound.right);
      bottom = max(bottom, bound.bottom);
    });

    return Rect.fromLTRB(left, top, right, bottom);
  }

  static Path scalePath(Path data, double ratio) {
    final transform = Matrix4.identity();
    transform.scale(ratio, ratio);

    return data.transform(transform.storage);
  }

  static List<Path> scalePaths(List<Path> data, double ratio) {
    final output = <Path>[];

    data.forEach((path) => output.add(scalePath(path, ratio)));

    return output;
  }

  static List<Path> translatePaths(List<Path> data, Offset location) {
    final output = <Path>[];

    final transform = Matrix4.identity();
    transform.translate(location.dx, location.dy);

    data.forEach((path) => output.add(path.transform(transform.storage)));

    return output;
  }

  static List<Path> parseDrawable(DrawableParent root) {
    final list = <Path>[];

    _parseDrawableRoot(root, list);

    return PathUtil.translatePaths(list, -PathUtil.pathBounds(list).topLeft);
  }

  static _parseDrawableRoot(DrawableParent root, List<Path> output) {
    if (root.children != null) {
      root.children!.forEach((drawable) {
        if (drawable is DrawableShape) {
          output.add(drawable.path);
        } else if (drawable is DrawableParent) {
          _parseDrawableRoot(drawable, output);
        }
      });
    }
  }

  static Size getDrawableSize(DrawableRoot root) => root.viewport.size;

  static Path toShapePath(List<CubicLine> lines, double size, double? maxSize) {
    assert(lines.length > 0);

    if (lines.length == 1) {
      final line = lines[0];
      if (line.isDot) {
        //TODO: return null or create circle ?
        return Path()
          ..start(line.start)
          ..line(line.end);
      }

      return line.toShape(size, maxSize!);
    }

    final path = Path();

    final firstLine = lines.first;
    path.start(firstLine.start + firstLine.cpsUp(size, maxSize!));

    for (int i = 0; i < lines.length; i++) {
      final line = lines[i];
      final d1 = line.cpsUp(size, maxSize);
      final d2 = line.cpeUp(size, maxSize);

      path.cubic(line.cpStart + d1, line.cpEnd + d2, line.end + d2);
    }

    final lastLine = lines.last;
    path.line(lastLine.end + lastLine.cpeDown(size, maxSize));

    for (int i = lines.length - 1; i > -1; i--) {
      final line = lines[i];
      final d3 = line.cpeDown(size, maxSize);
      final d4 = line.cpsDown(size, maxSize);

      path.cubic(line.cpEnd + d3, line.cpStart + d4, line.start + d4);
    }

    path.close();

    return path;
  }

  static Path toLinePath(List<CubicLine> lines) {
    assert(lines.length > 0);

    final path = Path()..start(lines[0]);

    lines.forEach((line) => path.cubic(line.cpStart, line.cpEnd, line.end));

    return path;
  }
}


lib/src/signature_view.dart
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';
import 'package:flutter_svg/flutter_svg.dart';
import 'package:flutter_svg/parser.dart';

import '../story_painter.dart';
import 'utils.dart';

class StoryPainter extends StatelessWidget {
  final StoryPainterControl control;

  StoryPainter({Key? key, required this.control}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return RepaintBoundary(
      key: control.painterKey,
      child: Stack(
        children: [
          StoryPainterPaint(
            control: control,
            onSize: control.notifyDimension,
          ),
          GestureDetector(
            onLongPressStart: (args) => control.startPath(args.localPosition),
            onLongPressEnd: (args) => control.closePath(),
            onLongPressMoveUpdate: (args) =>
                control.alterPath(args.localPosition),
            onTapUp: (args) => control.closePath(),
            //onTapCancel: () => control.closePath(),
            onTapDown: (args) => control.startPath(args.localPosition),
            onScaleStart: (args) => control.startPath(args.localFocalPoint),
            onScaleUpdate: (args) => control.alterPath(args.localFocalPoint),
            onScaleEnd: (args) => control.closePath(),
          ),
        ],
      ),
    );
  }
}

class StoryPainterView extends StatelessWidget {
  final Drawable? data;
  final Color? color;
  final double Function(double? width)? strokeWidth;
  final EdgeInsets? padding;
  final Widget? placeholder;

  const StoryPainterView({
    Key? key,
    required this.data,
    this.color,
    this.strokeWidth,
    this.padding,
    this.placeholder,
  }) : super(key: key);

  static _StoryPainterViewSvg svg({
    Key? key,
    required String data,
    Color? color,
    double Function(double? width)? strokeWidth,
    EdgeInsets? padding,
    Widget? placeholder,
  }) =>
      _StoryPainterViewSvg(
        key: key,
        data: data,
        color: color,
        strokeWidth: strokeWidth,
        padding: padding,
        placeholder: placeholder,
      );

  @override
  Widget build(BuildContext context) {
    if (data == null) {
      return placeholder ?? Container(color: Colors.transparent);
    }

    return Padding(
      padding: padding ?? EdgeInsets.all(8.0),
      child: FittedBox(
        fit: BoxFit.contain,
        alignment: Alignment.center,
        child: SizedBox.fromSize(
          size: PathUtil.getDrawableSize(data as DrawableRoot),
          child: CustomPaint(
            painter: DrawableSignaturePainter(
              drawable: data as DrawableParent,
              color: color,
              strokeWidth: strokeWidth,
            ),
          ),
        ),
      ),
    );
  }
}

class _StoryPainterViewSvg extends StatefulWidget {
  final String data;
  final Color? color;
  final double Function(double? width)? strokeWidth;
  final EdgeInsets? padding;
  final Widget? placeholder;

  const _StoryPainterViewSvg({
    Key? key,
    required this.data,
    this.color,
    this.strokeWidth,
    this.padding,
    this.placeholder,
  }) : super(key: key);

  @override
  _StoryPainterViewSvgState createState() => _StoryPainterViewSvgState();
}

class _StoryPainterViewSvgState extends State<_StoryPainterViewSvg> {
  DrawableParent? drawable;

  @override
  void initState() {
    super.initState();

    _parseData(widget.data);
  }

  void _parseData(String data) async {
    if (data == null) {
      drawable = null;
    } else {
      final parser = SvgParser();
      drawable = await parser.parse(data);
    }

    setState(() {});
  }

  @override
  void didUpdateWidget(_StoryPainterViewSvg oldWidget) {
    super.didUpdateWidget(oldWidget);

    if (oldWidget.data != widget.data) {
      if (drawable != null) {
        setState(() {
          drawable = null;
        });
      }

      _parseData(widget.data);
    }
  }

  @override
  Widget build(BuildContext context) => StoryPainterView(
        data: drawable,
        color: widget.color,
        strokeWidth: widget.strokeWidth,
        padding: widget.padding,
        placeholder: widget.placeholder,
      );
}

class _SinglePanGestureRecognizer extends PanGestureRecognizer {
  _SinglePanGestureRecognizer({Object? debugOwner})
      : super(debugOwner: debugOwner);

  bool isDown = false;

  @override
  void addAllowedPointer(PointerEvent event) {
    if (isDown) {
      return;
    }

    isDown = true;
    super.addAllowedPointer(event as PointerDownEvent);
  }

  @override
  void handleEvent(PointerEvent event) {
    super.handleEvent(event);

    if (!event.down) {
      isDown = false;
    }
  }
}


lib/src/signature_paint.dart
import 'package:flutter/material.dart';

import '../story_painter.dart';

class StoryPainterPaint extends StatefulWidget {
  final StoryPainterControl? control;
  final bool Function(Size size)? onSize;

  const StoryPainterPaint({
    Key? key,
    this.control,
    this.onSize,
  }) : super(key: key);

  @override
  StoryPainterPaintState createState() => StoryPainterPaintState();
}

class StoryPainterPaintState extends State<StoryPainterPaint> {
  List<SinglePath> pathWidgets = <SinglePath>[];
  @override
  void initState() {
    super.initState();
    widget.control!.paths.forEach((_path) {
      pathWidgets.add(SinglePath(
        path: _path,
        onSize: widget.onSize,
        type: _path!.type,
      ));
    });
    widget.control!.pageState = this;
  }

  void add() {
    pathWidgets.add(
      SinglePath(
        key: ObjectKey(widget.control!.paths.last!.id),
        path: widget.control!.paths.last,
        onSize: widget.onSize,
        type: widget.control!.type,
      ),
    );
    refreshState();
  }

  void update() {
    pathWidgets.last.path?.pathState?.refreshState();
  }

  void remove() {
    pathWidgets.removeLast();
    refreshState();
  }

  @override
  Widget build(BuildContext context) {
    return Stack(
      children: pathWidgets,
    );
  }

  void refreshState() {
    if (this.mounted) {
      setState(() {});
    }
  }
}

class SinglePath extends StatefulWidget {
  final CubicPath? path;
  final PainterDrawType? type;
  final bool Function(Size size)? onSize;

  const SinglePath({Key? key, this.type, this.onSize, this.path})
      : super(key: key);

  @override
  SinglePathState createState() => SinglePathState();
}

class SinglePathState extends State<SinglePath> {
  @override
  void initState() {
    super.initState();
    widget.path!.pathState = this;
  }

  @override
  Widget build(BuildContext context) {
    return RepaintBoundary(
      child: CustomPaint(
        isComplex: true,
        willChange: false,
        painter: PathSignaturePainter(
          path: widget.path!,
          onSize: widget.onSize,
        ),
      ),
    );
  }

  void refreshState() {
    if (this.mounted) {
      setState(() {});
    }
  }
}


lib/src/signature_painter.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:flutter_svg/flutter_svg.dart';

import '../story_painter.dart';
import 'utils.dart';

enum PainterDrawType {
  line,
  arc,
  shape,
}

class PathSignaturePainter extends CustomPainter {
  final CubicPath path;
  final bool Function(Size size)? onSize;

  Paint get strokePaint => Paint()
    ..style = PaintingStyle.stroke
    ..strokeCap = StrokeCap.round
    ..strokeJoin = StrokeJoin.round;

  Paint get fillPaint => Paint()..strokeWidth = 0.0;
  double a = 4;
  double b = 0.2;

  double multiplier() => ((a * pow(1 - b, path.width - 2)) + 1).toDouble();

  double? _maxWidth() {
    if (maxWCache.containsKey(path.width)) {
      return maxWCache[path.width];
    }
    maxWCache[path.width] = path.width * multiplier();
    return maxWCache[path.width];
  }

  PathSignaturePainter({
    required this.path,
    this.onSize,
  }) : assert(path != null);

  @override
  void paint(Canvas canvas, Size size) {
    //TODO: move to widget/state
    if (onSize != null) {
      if (onSize!(size)) {
        return;
      }
    }

    if (path.lines.isEmpty) {
      return;
    }

    switch (path.type) {
      case PainterDrawType.line:
        final paint = strokePaint;

        double _width = path.width;
        canvas.drawPath(
            PathUtil.toLinePath(path.lines),
            paint
              ..color = path.color
              ..strokeWidth = _width);
        break;
      case PainterDrawType.arc:
        final paint = strokePaint;

        double _width = path.width;
        path.arcs.forEach((arc) {
          paint.strokeWidth = _width + (_maxWidth()! - _width) * arc.size;
          canvas.drawPath(
              arc.path,
              paint
                ..color = path.color
                ..strokeWidth = _width);
        });
        break;
      case PainterDrawType.shape:
        final paint = fillPaint;

        double _width = path.width;
        if (path.isFilled) {
          if (path.isDot) {
            canvas.drawCircle(
              path.lines[0].start,
              path.lines[0].startRadius(_width, _maxWidth()!),
              paint..color = path.color,
            );
          } else {
            canvas.drawPath(
              PathUtil.toShapePath(path.lines, _width, _maxWidth()),
              paint..color = path.color,
            );

            final first = path.lines.first;
            final last = path.lines.last;

            canvas.drawCircle(
                first.start,
                first.startRadius(_width, _maxWidth()!),
                paint
                  ..color = path.color
                  ..strokeWidth = path.width);
            canvas.drawCircle(
                last.end,
                last.endRadius(_width, _maxWidth()!),
                paint
                  ..color = path.color
                  ..strokeWidth = path.width);
          }
        }

        break;
    }
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) {
    return true;
  }
}

class DrawableSignaturePainter extends CustomPainter {
  final DrawableParent drawable;
  final Color? color;
  final double Function(double? width)? strokeWidth;

  DrawableSignaturePainter({
    required this.drawable,
    this.color,
    this.strokeWidth,
  }) : assert(drawable != null);

  @override
  void paint(Canvas canvas, Size size) {
    _draw(
      drawable,
      canvas,
      Paint()
        ..color = color ?? Colors.black
        ..strokeCap = StrokeCap.round
        ..strokeJoin = StrokeJoin.round,
    );
  }

  void _draw(DrawableParent root, Canvas canvas, Paint paint) {
    if (root.children != null) {
      root.children!.forEach((drawable) {
        if (drawable is DrawableShape) {
          final stroke = drawable.style.stroke;
          final fill = drawable.style.fill;

          if (fill != null && !DrawablePaint.isEmpty(fill)) {
            paint.style = PaintingStyle.fill;
            if (color == null && fill.color != null) {
              paint.color = fill.color!;
            }
          } else if (stroke != null && !DrawablePaint.isEmpty(stroke)) {
            paint.style = PaintingStyle.stroke;

            if (color == null && stroke.color != null) {
              paint.color = stroke.color!;
            }

            if (stroke.strokeWidth != null) {
              if (strokeWidth != null) {
                paint.strokeWidth = strokeWidth!(stroke.strokeWidth);
              } else {
                paint.strokeWidth = stroke.strokeWidth!;
              }
            }
          }

          canvas.drawPath(drawable.path, paint);
        } else if (drawable is DrawableParent) {
          _draw(drawable, canvas, paint);
        }
      });
    }
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) {
    return true;
  }
}

class DebugSignaturePainterCP extends CustomPainter {
  final StoryPainterControl control;
  final bool cp;
  final bool cpStart;
  final bool cpEnd;
  final bool dot;
  final Color color;

  DebugSignaturePainterCP({
    required this.control,
    this.cp: false,
    this.cpStart: true,
    this.cpEnd: true,
    this.dot: true,
    this.color: Colors.red,
  });

  @override
  void paint(Canvas canvas, Size size) {
    final paint = Paint()
      ..color = color
      ..style = PaintingStyle.stroke
      ..strokeCap = StrokeCap.round
      ..strokeJoin = StrokeJoin.round
      ..strokeWidth = 1.0;

    control.lines.forEach((line) {
      if (cpStart) {
        canvas.drawLine(line.start, line.cpStart, paint);
        if (dot) {
          canvas.drawCircle(line.cpStart, 1.0, paint);
          canvas.drawCircle(line.start, 1.0, paint);
        }
      } else if (cp) {
        canvas.drawCircle(line.cpStart, 1.0, paint);
      }

      if (cpEnd) {
        canvas.drawLine(line.end, line.cpEnd, paint);
        if (dot) {
          canvas.drawCircle(line.cpEnd, 1.0, paint);
        }
      }
    });
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) {
    return true;
  }
}


example/lib/main.dart
import 'dart:typed_data';

import 'package:fast_color_picker/fast_color_picker.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:story_painter/story_painter.dart';

import 'dart:ui' as ui;

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Home(),
    );
  }
}

class Home extends StatefulWidget {
  @override
  _HomeState createState() => _HomeState();
}

class _HomeState extends State<Home> {
  StoryPainterControl painterControl = StoryPainterControl(
    type: PainterDrawType.shape,
    threshold: 3.0,
    smoothRatio: 0.65,
    velocityRange: 2.0,
    color: Colors.black,
    width: 8,
    onDrawStart: () {},
    onDrawEnd: () {},
  );
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('story_painter'),
        actions: [
          IconButton(
            icon: Icon(Icons.undo),
            onPressed: () async {
              painterControl.undo();
            },
          ),
          IconButton(
            icon: Icon(Icons.save),
            onPressed: () async {
              ui.Image image = await painterControl.toImage(pixelRatio: 3.0);
              ByteData? byteData =
                  await image.toByteData(format: ui.ImageByteFormat.png);
              var pngBytes = byteData?.buffer.asUint8List();
              if (pngBytes == null) {
                return;
              }
              Navigator.of(context).push(
                MaterialPageRoute(
                  builder: (_) => OnlyImage(
                    imageData: pngBytes,
                  ),
                ),
              );
            },
          ),
        ],
      ),
      body: Stack(
        alignment: Alignment.bottomCenter,
        children: [
          StoryPainter(
            control: painterControl,
          ),
          Column(
            mainAxisAlignment: MainAxisAlignment.end,
            children: [
              Slider(
                value: painterControl.width,
                onChanged: (width) {
                  painterControl.setWidth(width);
                  setState(() {});
                },
                max: 30,
                min: 5,
              ),
              FastColorPicker(
                selectedColor: painterControl.color,
                onColorSelected: (color) {
                  painterControl.setColor(color);
                  setState(() {});
                },
              ),
            ],
          )
        ],
      ),
    );
  }
}

class OnlyImage extends StatelessWidget {
  final Uint8List imageData;

  const OnlyImage({Key? key, required this.imageData}) : super(key: key);
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('image'),
      ),
      body: Image.memory(
        imageData,
      ),
    );
  }
}
